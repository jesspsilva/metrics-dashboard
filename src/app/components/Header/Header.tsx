import styled from "styled-components";
import { SearchSelect, SearchSelectItem } from "@tremor/react";

const HeaderStyled = styled.header`
  display: flex;
  justify-content: space-between;
  align-items: center;

  @media (max-width: 768px) {
    flex-direction: column;
    align-items: flex-start;
    padding: 20px 0 40px;

    > * {
      width: 100%;
    }
  }
`;

const H1 = styled.h1`
  font-size: 2rem;
  color: var(--dark-blue);

  @media (max-width: 768px) {
    padding: 0 0 20px;
  }
`;

const SelectWrapper = styled.div`
  display: flex;
  align-items: center;
  gap: 14px;

  @media (max-width: 768px) {
    flex-direction: column;
    align-items: flex-start;
  }

  .select-category {
    width: auto;
    min-width: 300px;
    text-transform: capitalize;

    input {
      text-transform: capitalize;
    }
  }
`;

const InputLabelStyled = styled.p`
	color: var(--neutral-dark);
	font-weight: 500;
}`;

export default function Header({
  categories,
  selectedCategory,
  onChange,
}: {
  categories: string[];
  selectedCategory: string;
  onChange: (v: string) => void;
}) {
  return (
    <HeaderStyled>
      <H1>Metrics</H1>
      <SelectWrapper>
        <InputLabelStyled id="category-label">
          Select a category
        </InputLabelStyled>
        <SearchSelect
          value={selectedCategory}
          onValueChange={(v) => onChange(v)}
          className="select-category"
          data-testid="category-select"
          defaultValue={categories[0]}
          enableClear={false}
        >
          {categories.map((category, index) => {
            return (
              <SearchSelectItem
                key={index}
                value={category}
                data-testid="category-option"
              >
                {category}
              </SearchSelectItem>
            );
          })}
        </SearchSelect>
      </SelectWrapper>
    </HeaderStyled>
  );
}
