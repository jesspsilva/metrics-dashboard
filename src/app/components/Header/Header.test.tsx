import "@testing-library/jest-dom";
import { render, screen, within } from "@testing-library/react";
import Header from "./Header";
import userEvent from "@testing-library/user-event";

const defaultTestProps = {
  categories: ["efficiency", "shift"],
  selectedCategory: "efficiency",
  onChange: () => {},
};

const renderComponent = (props = {}) => {
  const mergedProps = { ...defaultTestProps, ...props };
  return render(<Header {...mergedProps} />);
};

describe("Header", () => {
  it("should render correctly", () => {
    const { asFragment } = renderComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should display label", async () => {
    renderComponent();

    expect(await screen.findByText("Select a category")).toBeInTheDocument();
  });

  it("should render the correct number of categories", async () => {
    renderComponent();

    const select = await screen.findByTestId("category-select");
    const dropdown = within(select).getByRole("combobox");
    await userEvent.click(dropdown);
    expect(await within(select).findAllByRole("option")).toHaveLength(
      defaultTestProps.categories.length,
    );
  });

  describe("when selecting a category", () => {
    it("should call onChange", async () => {
      const onChange = jest.fn();
      renderComponent({ onChange });

      const select = await screen.findByTestId("category-select");
      const dropdown = within(select).getByRole("combobox");

      await userEvent.click(dropdown);

      const shiftOption = await within(select).findByText("shift");
      await userEvent.click(shiftOption);

      expect(onChange).toHaveBeenCalled();
    });
  });
});
