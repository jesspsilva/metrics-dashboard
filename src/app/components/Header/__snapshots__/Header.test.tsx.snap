// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render correctly 1`] = `
<DocumentFragment>
  <header
    class="sc-977317a-0 jDXbcE"
  >
    <h1
      class="sc-977317a-1 fKZkmD"
    >
      Metrics
    </h1>
    <div
      class="sc-977317a-2 ksfZbF"
    >
      <p
        class="sc-977317a-3 hGopXJ"
        id="category-label"
      >
        Select a category
      </p>
      <div
        class="w-full min-w-[10rem] text-tremor-default select-category"
      >
        <div
          class="relative"
        >
          <select
            class="h-full w-full absolute left-0 top-0 -z-10 opacity-0"
            title="search-select-hidden"
          >
            <option
              class="hidden"
              disabled=""
              hidden=""
              value=""
            >
              Select...
            </option>
            <option
              class="hidden"
              value="efficiency"
            >
              efficiency
            </option>
            <option
              class="hidden"
              value="shift"
            >
              shift
            </option>
          </select>
          <div
            data-headlessui-state=""
            data-testid="category-select"
          >
            <button
              aria-expanded="false"
              aria-haspopup="listbox"
              class="w-full"
              data-headlessui-state=""
              id="headlessui-combobox-button-:r0:"
              tabindex="-1"
              type="button"
            >
              <input
                aria-autocomplete="list"
                aria-expanded="false"
                autocomplete="off"
                class="w-full outline-none text-left whitespace-nowrap truncate rounded-tremor-default focus:ring-2 transition duration-100 text-tremor-default pr-14 border py-2 shadow-tremor-input focus:border-tremor-brand-subtle focus:ring-tremor-brand-muted dark:shadow-dark-tremor-input dark:focus:border-dark-tremor-brand-subtle dark:focus:ring-dark-tremor-brand-muted pl-3 placeholder:text-tremor-content dark:placeholder:text-tremor-content bg-tremor-background dark:bg-dark-tremor-background hover:bg-tremor-background-muted dark:hover:bg-dark-tremor-background-muted text-tremor-content-emphasis dark:text-dark-tremor-content-emphasis border-tremor-border dark:border-dark-tremor-border"
                data-headlessui-state=""
                id="headlessui-combobox-input-:r1:"
                placeholder="Select..."
                role="combobox"
                type="text"
                value="efficiency"
              />
              <div
                class="absolute inset-y-0 right-0 flex items-center pr-2.5"
              >
                <svg
                  class="tremor-SearchSelect-arrowDownIcon flex-none h-5 w-5 !text-tremor-content-subtle !dark:text-dark-tremor-content-subtle"
                  fill="currentColor"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M11.9999 13.1714L16.9497 8.22168L18.3639 9.63589L11.9999 15.9999L5.63599 9.63589L7.0502 8.22168L11.9999 13.1714Z"
                  />
                </svg>
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
